data/dataset.py

```diff
@@ -0,0 +1,144 @@
+import torch
+import math
+from torch_geometric.data import Data, HeteroData, InMemoryDataset
+from torch import Tensor
+from typing import Union, Optional
+
+
+def get_negative_edges(
+    filter_ids: Tensor,
+    potential_edges: Tensor,
+    num_negative_edges: int = 10,
+) -> Tensor:
+    # Get the biggest value available in articles (potential edges to sample from)
+    id_max = torch.max(potential_edges, dim=1)[0][1]
+
+    # Create list of potential negative edges, filter out positive edges
+    combined = torch.cat((torch.range(start=0, end=id_max), filter_ids))
```

Comment:
I think this can be done once almost like a constant, right? I assume it takes some time, and the number of articles will not change throughout the lifetime of the model

Reply:
To be honest if you think about it: everything can be precalculated in the beginning.

Reply:
but it will mean that we are loading in huge data to the memory. @itchingpixels so why would we precalculate only the negative edges?

Reply:
is it a lot of memory? I thought this is just one tensor with length of the num_articles? or did I misunderstood something?

Reply:
No-no, I mean if we want to precalculate things, why only the negative edges. Or why exactly the negative edges and not anything else? ie we could precalculate all the HeteroData buckets per user during initalization.